jax/lib/xla_bridge.py

```diff
@@ -292,6 +292,41 @@ def constant(builder, py_val, canonicalize_types=True):
   else:
     raise TypeError("No constant handler for type: {}".format(py_type))
 
+# XLA supports arbitrary tuple nesting, but JAX only uses one level of tupling
+# (and our type checkers don't support recursive types).
+SpatialSharding = Union[Sequence[int], None, Tuple[Sequence[int], None]]
+
+def _sharding_to_proto(sharding: SpatialSharding):
```

Comment:
I cannot tell what this code is supposed to do, since one would need to understand the OpSharding proto. I wonder if it is easy to explain the requirements on the input, e.g., if the input is a sequence of integers, it must have the same length as the data rank, and the product is the number of devices?

Reply:
I expanded the comment on the SpatialSharding type annotation since that's used in multiple functions here, and linked to the OpSharding proto here (which could probably be more clearly documented itself...). Let me know if specific aspects are still confusing.